### PR TITLE
add jsonpath to kubectl

### PR DESCRIPTION
**Pull request by daizuozhuo**
@brendandburns I add jsonpath as an output option for kubectl.  And I think I should make a page to tell people the jsonpath syntax. Where should I put it?

### CODE REVIEW

File: pkg/kubectl/resource_printer.go [hunk 1]
```diff
@@ -1174,3 +1184,26 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
...
+
+// JSONPathPrinter is an implementation of ResourcePrinter which formats data with jsonpath expression.
+type JSONPathPrinter struct {
+	rawTemplate string
+	template    *jsonpath.JSONPath
+}
+
+func NewJSONPathPrinter(tmpl []byte) (*JSONPathPrinter, error) {
+	j := jsonpath.New("out")
+	err := j.Parse(string(tmpl))
```
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]:
```go
if err := j.Parse(string(tmpl)); err != nil {
   return nil, err
}
```

File: pkg/kubectl/resource_printer.go [hunk 2]
```diff
@@ -1174,3 +1184,26 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
...
 	return v, false
 }
+
+// JSONPathPrinter is an implementation of ResourcePrinter which formats data with jsonpath expression.
+type JSONPathPrinter struct {
+	rawTemplate string
+	template    *jsonpath.JSONPath
+}
+
+func NewJSONPathPrinter(tmpl []byte) (*JSONPathPrinter, error) {
```
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: It's weird that you are casting to []byte above, and then back to string below, just make the type of `tmpl` be `string`

File: pkg/kubectl/resource_printer.go [hunk 3]
```diff
@@ -1174,3 +1184,26 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
 	}
 	return v, false
 }
+
+// JSONPathPrinter is an implementation of ResourcePrinter which formats data with jsonpath expression.
+type JSONPathPrinter struct {
+	rawTemplate string
```
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: given that rawTemplate is never used, I would remove it from this struct.

File: pkg/kubectl/resource_printer.go [hunk 4]
```diff
@@ -1174,3 +1184,26 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
...
+	template    *jsonpath.JSONPath
```
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: if you remove the rawTemplate member, then I would just make this an embedded field, instead of a named field.

https://golang.org/doc/effective_go.html#embedding

**Comment by brendandburns** [comment]
Small comments, basically LGTM.

You need to run `./hack/run-gendoc.sh` to regenerate kubectl docs.

**Comment by brendandburns** [comment]
I would add the jsonpath documentation to `docs/user-guide/...` and add a link in the kubectl help docs.

**Comment by derekwaynecarr** [comment]
Can something exercise this in ```hack/test-cmd.sh```?

**Comment by daizuozhuo** [comment]
@derekwaynecarr Yes, I will add some tests to `hack/test.cmd.sh`.

File: docs/user-guide/kubectl/kubectl_expose.md [hunk 5]
```diff
@@ -69,10 +69,11 @@ $ kubectl expose rc streamer --port=4100 --protocol=udp --name=video-stream
       --dry-run=false: If true, only print the object that would be sent, without creating it.
       --generator="service/v2": The name of the API generator to use. There are 2 generators: 'service/v1' and 'service/v2'. The only difference between them is that service port in v1 is named 'default', while it is left unnamed in v2. Default is 'service/v2'.
   -h, --help=false: help for expose
+  -j, --jsonpath="": JSONPath template string to use when -o=jsonpath. The jsonpath template is composed of jsonpath expressions enclosed by {}
```
Comment (thockin) [inline docs/user-guide/kubectl/kubectl_expose.md]: Is it too late to argue for -o jsonpath="template string" and -o template="template string".

Comment (bgrant0607) [inline docs/user-guide/kubectl/kubectl_expose.md]: Not too late. See:
https://github.com/GoogleCloudPlatform/kubernetes/pull/12220/files#r36381964

Comment (brendandburns) [inline docs/user-guide/kubectl/kubectl_expose.md]: that's fine with me too.

@daizuozhuo do you want to do a separate PR to do that?

Comment (bgrant0607) [inline docs/user-guide/kubectl/kubectl_expose.md]: If you want to do that as a separate PR, then in this one we should:

a) Keep the jsonpath output format
b) Drop the new jsonpath flag and reuse the existing template and templatefile flags, just interpreting them based on the requested output format.

Comment (daizuozhuo) [inline docs/user-guide/kubectl/kubectl_expose.md]: Yes, I would do this in a separate RP.

Comment (daizuozhuo) [inline docs/user-guide/kubectl/kubectl_expose.md]: @bgrant0607 Good Suggestions!

**Comment by bgrant0607** [comment]
The command-line args LGTM. Please address comments from @brendandburns and @derekwaynecarr.

**Comment by brendandburns** [comment]
LGTM.  Needs rebase due to the grand rename.

**Comment by daizuozhuo** [comment]
@brendandburns Rebased

**Comment by daizuozhuo** [comment]
@derekwaynecarr How to make hack/test-cmd.sh run? It reports error like:
```
/Users/daizuozhuo/goproj/src/github.com/GoogleCloudPlatform/kubernetes/_output/local/bin/darwin/amd64/kubelet --really-crash-for-testing=true --root-dir=/tmp/kubelet.31768 --cert-dir=/var/folders/81/n6gk7mk1279_b4x2vqdcp1680000gn/T/ --docker-endpoint=fake:// --hostname-override=127.0.0.1 --address=127.0.0.1 --port=10250 --healthz-port=10248
I0810 11:25:08.464685   31811 plugins.go:70] No cloud provider specified.
W0810 11:25:08.464920   31811 server.go:407] setting OOM scores is unsupported in this build
W0810 11:25:08.465017   31811 server.go:639] No api server defined - no events will be sent to API server.
failed to create kubelet: cAdvisor is unsupported in this build
!!! [0810 11:25:20] Timed out waiting for kubelet(masterless) to answer at http://127.0.0.1:10248/healthz; tried 25 waiting 0.5 between each
!!! Error in ./test-cmd.sh:121
  'return 1' exited with status 1
Call stack:
  1: ./test-cmd.sh:121 main(...)
```

**Comment by daizuozhuo** [comment]
Comments are addressed. By the way, I noticed that hack/test-cmd.sh is not working in Mac OS, So I tested it in Linux.

**Comment by daizuozhuo** [comment]
I will change it to `-o go-template=...` , `-o go-template-file=...` and `-o jsonpath=...` in another PR.

**Comment by daizuozhuo** [comment]
@brendandburns I added `jsonpath.md`. Please comment if it has any problem.

**Comment by daizuozhuo** [comment]
@brendandburns @bgrant0607 When will it be merged?

File: pkg/kubectl/resource_printer_test.go [hunk 6]
```diff
@@ -164,6 +164,23 @@ func TestPrintTemplate(t *testing.T) {
 	}
 }
 
+func TestPrintJSONPath(t *testing.T) {
+	buf := bytes.NewBuffer([]byte{})
+	printer, found, err := GetPrinter("jsonpath", "{.metadata.name}")
```
Comment (bgrant0607) [inline pkg/kubectl/resource_printer_test.go]: I understand that this is patterned after TestPrintTemplate, but what I'd like to see is a table-driven test that exercises more cases than just a single field. 

An example of a table-driven test:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate_test.go#L52

The test could be expanded in a subsequent PR, though.

Comment (daizuozhuo) [inline pkg/kubectl/resource_printer_test.go]: Thanks. I changed it and moved several other test function into the table.

File: pkg/kubectl/resource_printer.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/resource_printer.go]: Please address this comment

**Comment by brendandburns** [comment]
I still see one [comment](https://github.com/kubernetes/kubernetes/pull/12202#discussion_r36352628) that hasn't been addressed, can you take a look at it?

**Comment by daizuozhuo** [comment]
@brendandburns I use rawTemplate to print debug message now.

File: pkg/kubectl/resource_printer.go [hunk 7]
```diff
@@ -1199,3 +1209,37 @@ func indirect(v reflect.Value) (rv reflect.Value, isNil bool) {
...
+	if err != nil {
+		return err
+	}
+	out := map[string]interface{}{}
+	if err := json.Unmarshal(data, &out); err != nil {
+		return err
+	}
+	if err = j.JSONPath.Execute(w, out); err != nil {
+		fmt.Fprintf(w, "Error executing template: %v\n", err)
+		fmt.Fprintf(w, "template was:\n\t%v\n", j.rawTemplate)
```
Comment (daizuozhuo) [inline pkg/kubectl/resource_printer.go]: @brendandburns print rawTemplate here

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

**Comment by brendandburns** [comment]
LGTM

**Comment by saad-ali** [comment]
Rebase needed

**Comment by daizuozhuo** [comment]
@saad-ali Rebased, thanks.

**Comment by smarterclayton** [comment]
Tests are flaking:

```
ok k8s.io/kubernetes/pkg/util/flushwriter	1.041s	coverage: 100.0% of statements
--- FAIL: TestKubenates (0.02 seconds)
jsonpath_test.go:47: in recursive name, expect to get "127.0.0.1 127.0.0.2 myself e2e", got "myself e2e 127.0.0.1 127.0.0.2"
FAIL
coverage: 90.2% of statements
FAIL k8s.io/kubernetes/pkg/util/jsonpath 0.129s
ok k8s.io/kubernetes/pkg/util/mount 1.033s coverage: 25.5% of statements
ok k8s.io/kubernetes/pkg/util/config 1.069s coverage: 92.9% of statements
ok k8s.io/kubernetes/pkg/storage	1.284s	coverage: 89.0% of statements
```

**Comment by daizuozhuo** [comment]
@smarterclayton Since the JSON input is a map, JSONPath may iterate it in unknown order in different machines. Sorry, I will change the test method in next PR.
